diff --git a/src/gallium/drivers/i915/i915_context.c b/src/gallium/drivers/i915/i915_context.c
index 28ff40a..72b62f8 100644
--- a/src/gallium/drivers/i915/i915_context.c
+++ b/src/gallium/drivers/i915/i915_context.c
@@ -171,8 +171,8 @@ i915_create_context(struct pipe_screen *screen, void *priv)
    i915_init_resource_functions(i915);
    i915_init_query_functions(i915);
 
-   draw_install_aaline_stage(i915->draw, &i915->base);
-   draw_install_aapoint_stage(i915->draw, &i915->base);
+//   draw_install_aaline_stage(i915->draw, &i915->base);
+//   draw_install_aapoint_stage(i915->draw, &i915->base);
    draw_enable_point_sprites(i915->draw, TRUE);
 
    /* augmented draw pipeline clobbers state functions */
