--- xorg-server-1.6.0/hw/xfree86/common/xf86AutoConfig.c~	2009-03-26 22:10:45.000000000 -0700
+++ xorg-server-1.6.0/hw/xfree86/common/xf86AutoConfig.c	2009-03-26 22:10:45.000000000 -0700
@@ -171,6 +171,9 @@
 	case 0x8086:
 	    if ((dev->device_id == 0x00d1) || (dev->device_id == 0x7800)) {
 		driverList[0] = "i740";
+	    } if ((dev->device_id == 0x8109) ||  (dev->device_id == 0x8108) ||  (dev->device_id == 0x4102) ) {
+	        driverList[0] = "psb";
+		driverList[1] = "psb_drv";
 	    } else {
 		driverList[0] = "intel";
 		driverList[1] = "i810";
